File: 0018-4sum/0018-4sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    let n = nums.length
    let answer = []
    
    let compareFn = (a,b) => {
        if (a < b){
            return -1
        }
        else if (a > b){
            return 1
        }
        else{
            return 0
        }
    }
    
    nums.sort(compareFn)
    let i = 0
    
    while (i < n- 2){
        let j = i + 1
        
        while (j < n - 1){
            let k = j + 1
            let l = n - 1
            
            while (k < l){
                let sum_ = nums[i] + nums[j] + nums[k] + nums[l]
                
                if (sum_ === target){
                    answer.push([nums[i], nums[j], nums[k], nums[l]])
                    
                    while (nums[k] === nums[k+1]){
                        k += 1
                    }
                
                    while (nums[l] === nums[l-1]){
                        l -= 1
                    }
                
                    k += 1
                    l -= 1
                }
                else if (sum_ > target){
                    l -= 1
                }
                else {
                    k += 1
                }
            }
            
            while (j < n - 1 && nums[j] === nums[j+1]){
                j += 1
            }
            
            j += 1
            
        }
        
        while (i < n - 2 && nums[i] === nums[i+1]){
            i += 1
        }
        
        i += 1
    }
    
    return answer
};

File: 0018-4sum/0018-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        
        n = len(nums)
        ans = list()
        nums.sort()
        for i in range(n):

            if i != 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i+1, n):

                if j > i+ 1 and nums[j] == nums[j - 1]:
                    continue

                left = j + 1
                right = n - 1

                while left < right:

                    total_sum = nums[i] + nums[j] + nums[left] + nums[right]

                    if total_sum < target:
                        left += 1

                    elif total_sum > target:
                        right -= 1

                    else:
                        ans.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1

                        while left < right and nums[left] == nums[left - 1]:
                            left += 1

                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1

        return ans

                